### ISSUE TITLE
No maxFuel information for drive_train_attributes

### ISSUE THREAD

**Issue by ronmichel** [bimmerconnected/bimmer_connected#129]
**Describe the issue**
```
 ERROR (MainThread) [homeassistant.components.sensor] bmw_connected_drive: Error on device update!
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 292, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 448, in async_device_update
    await self.hass.async_add_executor_job(self.update)
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/homeassistant/homeassistant/components/bmw_connected_drive/sensor.py", line 147, in update
    self._state = getattr(vehicle_state, self._attribute)
  File "/usr/local/lib/python3.7/site-packages/bimmer_connected/state.py", line 307, in __getattr__
    return self._attributes[item]
KeyError: 'maxFuel'
```

**Expected behavior**
Do not add  maxFuel to drivetrain attributes if not available

**Suspected code generating the error**
My car (530e) does have a combustion engine and hv battery but no maxFuel information.
See here where the issue seems to be: : https://github.com/bimmerconnected/bimmer_connected/blob/7bd82921209c492c2b9b5ca5c873077221b84adf/bimmer_connected/vehicle.py#L115-L116

**Additional**

`{"vehicleStatus":{"vin":"WBAXXX","mileage":8739,"updateReason":"CHARGING_DONE","updateTime":"2019-08-22T12:33:13+00:00","doorDriverFront":"CLOSED","doorDriverRear":"CLOSED","doorPassengerFront":"CLOSED","doorPassengerRear":"CLOSED","windowDriverFront":"CLOSED","windowDriverRear":"CLOSED","windowPassengerFront":"CLOSED","windowPassengerRear":"CLOSED","trunk":"CLOSED","rearWindow":"INVALID","hood":"CLOSED","doorLockState":"SECURED","parkingLight":"OFF","positionLight":"OFF","remainingFuel":34,"remainingRangeElectric":41,"remainingRangeElectricMls":25,"remainingRangeFuel":558,"remainingRangeFuelMls":346,"maxRangeElectric":43,"maxRangeElectricMls":26,"connectionStatus":"CONNECTED","chargingStatus":"FINISHED_FULLY_CHARGED","chargingLevelHv":100,"lastChargingEndReason":"CHARGING_GOAL_REACHED","lastChargingEndResult":"SUCCESS","position":{"lat":48.139305,"lon":11.548841,"heading":342,"status":"OK"},"publicKey":"XXX","internalDataTimeUTC":"2019-08-22T12:33:13","singleImmediateCharging":false,"chargingConnectionType":"CONDUCTIVE","chargingInductivePositioning":"NOT_POSITIONED","steering":"LH","vehicleCountry":"DE","checkControlMessages":[],"cbsData":[{"cbsType":"OIL","cbsState":"OK","cbsRemainingMileage":23000,"cbsDueDate":"2020-12","cbsDescription":"Nächster Service nach der angegebenen Fahrstrecke oder zum angegebenen Termin."},{"cbsType":"VEHICLE_CHECK","cbsState":"OK","cbsRemainingMileage":60000,"cbsDueDate":"2022-12","cbsDescription":"Nächste Sichtprüfung nach der angegebenen Fahrstrecke oder zum angegebenen Termin."},{"cbsType":"BRAKE_FLUID","cbsState":"OK","cbsDueDate":"2021-12","cbsDescription":"Nächster Wechsel spätestens zum angegebenen Termin."}],"DCS_CCH_Activation":"ACTIVE","DCS_CCH_Ongoing":false,"climateFunction":"SUPPORTED"}}`

**Comment by gerard33** [comment]
Thanks for the issue and the additional info. The code you mentioned is indeed causing this issue. Need to add an extra check to see if `maxFuel` is available.

**Comment by licheng5625** [comment]
same as chargingInductivePositioning
My car mini countryman phev has no this attribute.

**Comment by robthebold** [comment]
It looks like there are a number of cases where our logic of adding attributes breaks down wrt what the BMW servers return for a given model. Usually about high-voltage battery states and fuel status not conforming to the bimmer_connected model of what attributes should be available based on whether a car has an ICE or a traction battery.

I think we can improve the situation by checking if an attribute (key) exists in the status info we've pulled from the server before adding an attribute to the vehicle. I've written a function to check attributes before adding them to the list, but I'm in the middle of a different PR and I don't have the time to figure how to manage multiple PRs in git (vs. a real version control system) at once.

Basically, it would work like this:

`     def add_attributes(attrs, new_attrs: list) -> list:
        """
        Make sure newly added attributes are really in the vehicle status to avoid key error later
        camelCase the snake_case names before checking they exist
        """
        for new_attr in new_attrs:
            cc_attr = new_attr.split('_')[0] + ''.join(x.capitalize() or '_' for x in new_attr.split('_')[1:])
            if cc_attr in self.state.attributes:
                #If camelCase name is in status, then add the new attribute
                attrs += new_attr
        return attrs`

instead of doing a simple:

`attrs+=new_attrs`

list concatenation.

This should convert any snake_case attribute names to camelCase to check the data returned from the server to ensure that a given vehicle really has a particular attribute.

Sorry, that code above looks like garbage, don't really know how to pretty it up to resemble code :(. whatever.

**Comment by robthebold** [comment]
OK, it's daytime and I can paste in a much more readable snippet that tests whether an attribute is in the status returned from the vehicle before adding it to our list of available attributes. This exercise has made me curious about why some attributes are snake_case and some are camelCase, e.g. 'charging_time_remaining' vs 'chargingConnectionType'. Does this impact something on the HA side?

Anyway, the code snippet:

    def add_attributes(attrs, new_attrs: list) -> list:
        """
        Make sure newly added attributes are really in the vehicle status to avoid key error later
        camelCase the snake_case names before checking they exist
        """
        for new_attr in new_attrs:
            cc_attr = new_attr.split('_')[0] + ''.join(x.capitalize() or '_' for x in new_attr.split('_')[1:])
            if cc_attr in self.state.attributes:
                #If camelCase name is in status, then add the new attribute
                attrs += new_attr
        return attrs

**Comment by gerard33** [comment]
@robthebold thanks for the code snippet. Will have a look at this asap as this is apparently affecting a number of users.
The attributes in HA should be snake_case, but we also re-use some of the names from the BMW api and those are also camelCase. Changing this is a breaking change for HA if you use e.g. automations with the attributes, but that should be worth it in order to get in line with the standard method.

**Comment by gerard33** [comment]
Fixed in #148 and [HA PR](https://github.com/home-assistant/home-assistant/pull/31894) made.
